src/pybind/mgr/deepsea/module.py

```diff
@@ -0,0 +1,411 @@
+                # if node_filter was requested (DeepSea role should map to
+                # Orchestrator label concept)
+            })
+
+            # ['return'][0]['tag'] in the resonse JSON is what we need to match
+            # on when looking for the result event (e.g.: "salt/run/20181018074024331230")
+            self._all_completions["{}/ret".format(resp.json()['return'][0]['tag'])] = c
+
+            return c
+
+        # TODO: It's possilbe _do_request_with_login could throw an exception
+        # (e.g. salt-api down).  How to handle this?
+
+
+    def describe_service(self, service_type, service_id):
+        # Trivial implmenetation ignoring service_id (to test, try
+        # `ceph orchestrator service status mon ''`).  Also, it will lie horribly
+        # about OSDs.  This method is also faked up with 'select.minions' because
+        # again we don't have a runner that quite matches on the DeepSea side yet.
+
+        assert service_type in ("mon", "mgr", "osd", "mds", "rgw"), service_type + " unsupported"
+
+        def process_result(raw_event):
+            result = orchestrator.ServiceDescription()
+            raw_event = json.loads(raw_event)
+            if raw_event['data']['success']:
+                for hostname in raw_event["data"]["return"]:
+                    loc = orchestrator.ServiceLocation()
+                    loc.nodename = hostname
+                    # this next is probably correct for mon, mgr, mds, maybe rgw,
+                    # but will be very wrong for OSDs.
+                    loc.daemon_name = hostname.split('.')[0]
+                    result.locations.append(loc)
+            return result
+
+        with self._completion_lock:
+            c = DeepSeaReadCompletion(process_result)
+
+            role = "storage" if service_type == "osd" else service_type
+
+            resp = self._do_request_with_login("POST", data = {
+                "client": "runner_async",
+                "fun": "select.minions",
+                "cluster": "ceph",
+                "roles": role
+            })
+            self._all_completions["{}/ret".format(resp.json()['return'][0]['tag'])] = c
+
+            return c
+
+        # TODO: It's possilbe _do_request_with_login could throw an exception
+        # (e.g. salt-api down).  How to handle this?
+
+
+    def wait(self, completions):
+        incomplete = False
+
+        with self._completion_lock:
+            for c in completions:
+                if c.is_complete:
+                    continue
+                if not c.is_complete:
+                    # TODO: the job is in the bus, it should reach us eventually
+                    # unless something has gone wrong (mgr failed over, salt-api
+                    # died, etc.), in which case it's possible the job finished
+                    # but we never noticed the salt/run/$id/ret event.  Should
+                    # probably add the job ID (and start timestamp?) to the
```

Comment:
I think the job ID should be enough.

Will this help with a mgr failover though? Isn't this state gone in the failover case? Further down you mention potentially persisting completions. Would there be another way? This also seems like a problem all other orchestrator implementation will have .

Reply:
> I think the job ID should be enough.

I wondered if having a timestamp would let us not bother to query active jobs too often/too quickly, but may be overcomplicating/prematurely optimising, so, yeah, I agree, just the ID. We can always add more information later if it turns out we need it.

> Will this help with a mgr failover though?

Nope. It'll only help if mgr is still running, but the salt api died (or dropped connection for whatever reason)

> Isn't this state gone in the failover case?

Yep. Thinking about it again, maybe we don't care about persisting completions. I was wondering if a client of the orchestrator could still have a valid completion object, while mgr failed over to another node, then later the client passes that completion object to `wait()` on the new mgr instance. Except, the only things that can actually receive completion objects from the orchestrator module, are other mgr modules, which are necessarily running in the same mgr instance, and will _also_ die if mgr fails over, so everything just gets dropped on the floor.